### ISSUE TITLE
Support environments for shareable configs

### ISSUE THREAD

**Issue by Fer0x** [browserslist/browserslist#300]
I would like to keep my shareable config similarly to package.json field "browserslist", like this:

```js
module.exports = {
  development: ['last 2 chrome versions', 'last 2 firefox versions', 'last 2 edge versions'],
  production: [
    '>1%',
    'last 3 versions',
    'not dead',
    'not op_mini all',
    'Firefox ESR',
    'not ie < 11',
  ],
};
```

For now it's impossible because of error: `config exports not an array of queries `.

**Comment by ai** [comment]
Good idea.

I am not sure when I will fix it  (because most of tools doesn’t support environments, it is very advanced and rare feature). If you want it quicker, you can send PR.

**Comment by Fer0x** [comment]
@ai ok, I'll take a look at it soon.

**Comment by JamiesonRoberts** [comment]
@Fer0x Could the same pattern as eslint, prettier, and babel be used, where you can have use a `.browserlistrc.js` or `browserlist.config.js` file in your repo? This would keep it fitting the file patterns despite being different projects.

I know for example Prettier and I think esLint use Comsiconfig (https://github.com/davidtheclark/cosmiconfig) to handle finding the files. Happy to take a stab at it if you haven't done it yet.

**Comment by ai** [comment]
@JamiesonRoberts Browserslist can’t use JS configs and Comsiconfig, because Browserslist config should be able to be used by non-JS tools (like AutoprefixerRails has Ruby code to load Browserslist config)

**Comment by JamiesonRoberts** [comment]
@ai Maybe fair about cosmiconfig, but it should be possible to handle the array from a module.export the same as the package.json version. Ultimately as long as it exports an array would it not be possible to load from a .js file?

**Comment by ai** [comment]
@JamiesonRoberts if you want just `dev` vs `production` you can use Browserslist environments:
https://github.com/browserslist/browserslist#environments

Maybe it will be enough. Or you want completely dynamically queries?

**Comment by JamiesonRoberts** [comment]
@ai my preference is dynamic queries or setting things programmatically rather than relying on specific key in an object to match. That way I can retrieve the values in several different scenarios, rather than just simple ones like above.

**Comment by ai** [comment]
> my preference is dynamic queries or setting things programmatically

1. I do not like to add features without real use case. Often you just add features and nobody will use them later (but you still need to support them).
2. How Autoprefixer Rails (written on Ruby) will read this config?

**Comment by JamiesonRoberts** [comment]
@ai You definitely made me think on that this one, so here is my rational (and I'm sorry for the long read)

I work at an agency, which means we have hundreds of repos/projects. Over time we updates to projects happen and are deployed etc. While we do have a set list of browser we support based on the Browserlist spec, over time that does change as a core spec. Now updating hundreds of repos becomes incredibly burdensome and time consuming. The ability to extend a central config (for example the idea behind how esLint handles it) would be incredibly powerful. This could either be through an "extend" property, or something more powerful like being able to `module.export` as this issue outlines. 

This would give the power to extend that central configuration and then provide per project additions or overrides, but have a centrally updated config for the average project and the core rules. Saving time for updates and configurations.

One other reason for it but what I would consider as a more of a discussion around intepretation of spec is that `package.json` is technically a spec to handle the installation of dependancies, not manage the configurations of them. 

> A package.json file: lists the packages that your project depends on. allows you to specify the versions of a package that your project can use using semantic versioning rules.

Yes there is the ability to set a `.browserlistrc` file, it is not a prototypical JS file (eg it couldn't be lint-able for a JS project) which means that it then lives outside the JS ecosystem for a project and leaves a potential black-hole for verifying a project programmatically. I trust my developers on my team, but that doesn't mean that humans won't make a mistake accidentally and cause something to break like a deployment. And while Browserlist can be used in a Ruby project, there is technically nothing in the readme/docs for Browserlist stating that its goal is to be a cross language tool, all of the examples are javascript based ones.

As for **Autoprefixer Rails** If you look at the documentation for Autoprefixer Rails it already states that you can't use the `package.json` declaration for Browserlist, so to my mind this would be the same thing. That being said, it does leave the Ruby side of things at a disadvantage of not having dynamic declarations. On that side of things, my skill/knowledge is sadly lacking for Ruby, so I don't know how that could be accomplished, or if it could even be accomplished with a single file type or in a repeatable way.

**Comment by ai** [comment]
> I work at an agency, which means we have hundreds of repos/projects. Over time we updates to projects happen and are deployed etc. While we do have a set list of browser we support based on the Browserlist spec, over time that does change as a core spec. Now updating hundreds of repos becomes incredibly burdensome and time consuming. The ability to extend a central config (for example the idea behind how esLint handles it) would be incredibly powerful. 

Good use case. Don’t worry, the rule of all my projects is to always suggest a solution for every real use case. So we will find/create a solution for you.

Can you tell me, why combining queries will not work for your case:

```
extends @company/browserslist-config
not IE 10 # removing browser from sharable config selection
Firefox ESR # adding browser to previous selection
```

**Comment by JamiesonRoberts** [comment]
I'd be good with that approach. Would it have the ability to be pulled in via an NPM package or similar package manager?

EX: point it to a package name and then it would look there for either a `.browserlistrc` or the `package.json`? 

One thought would be how esLint handles it with a defined naming convention, they require your package to be named `eslint-config-{name}` so in this case it could be `browserslist-config-{name}` (docs https://eslint.org/docs/developer-guide/shareable-configs)

`.browserlistrc` file example
```
extends browserslist-config-company
not IE 10 # removing browser from sharable config selection
Firefox ESR # adding browser to previous selection
```

`package.json` example
```json
{
  "browserslist": [
    "extends browserslist-config-company",
    "> 1%",
    "not IE 10",
    "Firefox ESR"
  ]
}
```
I would argue for the `package.json` one for it to be more prototypical to how JSON works and clarity for the end user readability wise, it would be better as the following though. But I recognize that it would be a breaking change.
```json
{
  "browserslist": {
    "extends": "browserslist-config-company", 
    "rules" : [
      "> 1%",
      "not IE 10",
      "Firefox ESR"
    ]
  }
}
```
One advantage to the second json example is that your extends rule could be both a string or an array of strings, but I don't know about the usefulness of merging multiple extensions together.

**Comment by ai** [comment]
> Would it have the ability to be pulled in via an NPM package or similar package manager?

Sure, in `extends browserslist-config-company`, `browserslist-config-company` is npm package name.

> point it to a package name and then it would look there for either a .browserlistrc or the package.json?

Nope, it looks for `index.js`. Here is full doc https://github.com/browserslist/browserslist#shareable-configs

> One thought would be how esLint handles it with a defined naming convention, they require your package to be named eslint-config-{name}

Yeap, same in Browserslist. You _must_ to have `@NAME/browserslist-config` or `bowserslist-config-NAME` for security reasons.

> I would argue for the package.json one for it to be more prototypical to how JSON works and clarity for the end user readability wise

I do not want to have different syntaxes in `package.json`, `.browserslistrc` and `BROWSERSLIST` env variable.

> But I recognize that it would be a breaking change.

Yeap :D

**Comment by JamiesonRoberts** [comment]
@ai Well, that would definitely look to do it! I'll try it out and let you know if I have any problems with it! Thank you for your time and good discussion!

**Comment by sybrendotinga** [comment]
I was experimenting with implementing a shareable config for our company today. Because we also have two environments in our current setup I was looking into how to implement this. I ended up in this topic. For us this is a great use-case aswel. For now, I decided to only implement the production environment, that will also be used in our development environment. 

We would like to use this feature as @JamiesonRoberts describes it. It would be great to have a base configuration, but also have the option to override rules for a specific project.

**Comment by mbrevda** [comment]
This seems to be a feature request that has come up often ( #122, #186, #300, #408). Additionally, it would seem that the ability is already available when used via the API, simply by manually requiring/parsing the desired files - although this wouldn't work well with other tools (unless this was implemented there, too. i.e. to use a custom extend with babel one would need to use a bable.config.js and `require()` from there). It would be extremely useful to have this available out-of-the-box without needing to write custom code.

Regarding use cases, they all seem to share a common theme: larger orgs with many repos looking to share configs across projects.

The suggestion of Comsiconfig seems to be the simplest and can probably be done in a non-breaking way. Could we open a PR to see what this might look like?

**Comment by ai** [comment]
> Regarding use cases, they all seem to share a common theme: larger orgs with many repos looking to share configs across projects.

Why different shared configs is not a solution here? `@company/browserslist-config/banks` for instance.

> The suggestion of Comsiconfig seems to be the simplest and can probably be done in a non-breaking way.

`Comsiconfig` is overkill. We should be very careful with security (we have a lot of browsers where browserslist config is passed by the client). Using a big library, which we do not fully understand, is not a best-case here.

I prefer this syntax and will accept PR with it:

```js
module.exports = {
  production: ['last 2 version']
  development: ['last 1 version']
}
```

**Comment by mbrevda** [comment]
> Why different shared configs is not a solution here? @company/browserslist-config/banks for instance.

The naming convention is inflexible and doesn't work for everyone. As an example: imagain a company that wants to put all their configs in a single package - eslint, babel, webpack, etc. They cannot put the in their shared package as it doesn't begin with @company/browserslist. Now they need a second package just for the few lines lines of browserslist config.

> We should be very careful with security

I've seen this repeatedly across this repo with no clear explanation of what it means. Could you elaborate? Perhaps I'm missing something

> I prefer this syntax and will accept PR with it:

Some questions:

1. That seems like straight-up javascript. Is that correct? 
1. How would you like the file named? Should we follow the convention with `browserslist.config.js`? 
1. Could we just `require()` the file? 
1. What would the syntax be for a single env - `module.exports = ['last 2 version']`?

**Comment by ai** [comment]
@mbrevda 

> I've seen this repeatedly across this repo with no clear explanation of what it means. Could you elaborate? Perhaps I'm missing something

1. Browserslist call `require(configName)` and execute JS code in config’s package.
2. Some websites like [browserl.ist](https://browserl.ist/) accept Browserslist queries from users.
3. A user can pass something like `extends dev-tools/clean-everything.js` or `extends ./actions/delete-database.js`. It will force Browserslist to execute unexpected files in `node_modules` or project directory. It can be very dangerous.

> The naming convention is inflexible and doesn't work for everyone. 

We have this limit because of security concerns. But there is a `dangerousExtend` option to disable this check.

It should be a separated issue. We can think about adding `BROWSERSLIST_DANGEROUS_EXTEND` envirenment variable.

> That seems like straight-up javascript. Is that correct?

Yeap, simple ES5 JS without any builder. Only `node.js` allows Node.js API, `index.js` should be browser compatible.

> How would you like the file named? Should we follow the convention with browserslist.config.js?
> Could we just require() the file?

Browserslist already has shared config support. You do not need to change the shared config loading code.

https://github.com/browserslist/browserslist/blob/master/index.js#L1092-L1097
https://github.com/browserslist/browserslist/blob/master/node.js#L155-L165

Right now shared config should export only a string. In the PR we should also add `{ [env]: string }` support.

> What would the syntax be for a single env - module.exports = ['last 2 version']?

Yeap. It is what Browserslist expects from shared configs right now. We should keep it.

**Comment by mbrevda** [comment]
> Some websites like browserl.ist accept Browserslist queries from users.

If a website is trusting user input, I'm not sure there is much you can do. Additionally, what is the use case for allowing `extend` on browserl.ist? That seems like something which should be blocked. It certainly doesn't seem fair to penalize everyone else because browserl.ist might be built insecurely.

> You do not need to change the shared config loading code.

The PR will allow for native node `require()` behavior as another file parser. The results will be passed back to browerslist

**Comment by ai** [comment]
> Additionally, what is the use case for allowing extend on browserl.ist? That seems like something which should be blocked.

There is no sense, but the website developer could forget to add blocked for `extends`. Good API should not allow to shoot yourself in the foot.

> It certainly doesn't seem fair to penalize everyone else because browserl.ist might be built insecurely.

Let’s find limits and extra options to make it simpler for everyone.

> The PR will allow for native node require() behavior as another file parser. The results will be passed back to browerslist

Are you want to add JS format to Browserslist config?

1. We can’t do it, because of Browserslist config work in non-JS environments as well. For instance, [Autoprefixer Rails](https://github.com/ai/autoprefixer-rails) uses Ruby code to load files.
2. Dynamic files are bad for caching.

It is OK for rare `extends` query but too much for the config file.

**Comment by mbrevda** [comment]
> Browserslist config work in non-JS environments as well

There are many limitations of using browserslist in non-js environments. For example:

> Autoprefixer Rails doesn’t support these Autoprefixer features:
>
> Browsers in package.json.
> Custom browsers usage statistics.

Using a `.config.js` would be one more such limitation.

> caching

How often is this file read? What is the impact of not having it cached??

> Good API should not allow to shoot yourself in the foot.

browerslist is not a security tool, and should not be making security related decisions. It certainly shouldn't be FORCING security related decisions. These are implementational details best left to the implementing developer

> Let’s find limits and extra options to make it simpler for everyone.

Again, this package has one job: to parse a list of browsers. I'd be happy to add a PR that removes limits and allows greater flexibility in parsing the list of browsers.

In conclusion: it is my opinion that `dangerousExtend` should (be renamed `allowExtend` and) default to `true`. In the rare case of some extreme need that cannot be satisfied elsewhere, the dev can set it to `false`.

For now, I'll hold off on a PR until we can come to a consensus.

**Comment by ai** [comment]
> How often is this file read? What is the impact of not having it cached?

If we will use JS file, webpack/Parcel will not be sure that the content of the file was not changed.

> It certainly shouldn't be FORCING security related decisions.

I believe that every software should think about security.

Any your dependency can be used to hack the system. If a tool can potentially be used with non-trused user input, it is a security tool.

> For now, I'll hold off on a PR until we can come to a consensus.

This issue is about adding environments to shared config.

Removing mandatory `browserslist-config` prefix or adding JS-based config should be discussed in separated issue.

**Comment by ai** [comment]
Released at 4.12. Thanks to @yepninja for writing the code.

**Comment by timreichen** [comment]
> Removing mandatory `browserslist-config` prefix or adding JS-based config should be discussed in separated issue.

Created an issue [here](https://github.com/browserslist/browserslist/issues/495) and wrote a short article [here](https://gist.github.com/timreichen/87a894c94a4c2258b15962d1754e34a8) that is open for discussion about config files.
